rfcs/56-better-reflect.md

```diff
@@ -0,0 +1,261 @@
+
+The problem is not limited to `type_id`, but also extends to the `reflect_*`
+methods.
+
+```rust
+let a = MysteryType::new(); // suppose MysteryType implements Reflect
+let reflect_a: Box<dyn Reflect> = Box::new(a.clone()) as Box<dyn Reflect>;
+let dynamic_a: Box<dyn Reflect> = a.clone_value();
+let reflect_hash = reflect_a.reflect_hash();
+let dynamic_hash = dynamic_a.reflect_hash();
+// This more or less always panic if `MysteryType` implements `Hash`
+assert_eq!(reflect_hash, dynamic_hash)
+```
+
+
+## Proposition
+
+The problem stems from the duplicity of `dyn Reflect`. We want to rely on the
+underlying methods of the same thing to be unique.
+
+To solve this, we do not allow more than a single underlying type to "pass for"
+a concrete type in `Reflect`. This implies that the various `Dynamic***` types
+shouldn't implement `Reflect`.
+
+## User-facing explanation
+
+* `Reflect` maps 1-to-1 with an underlying type meaning that if a
+  `Box<dyn Reflect>` cannot be downcasted to `T`, you cannot build a `T` from
+  it.
+* There is a `ReflectProxy` enum you can use to construct any `Reflect` type
+  based on some dynamic structure. This can be used, for example, for
+  deserialization of types that implement `Reflect` but not `Deserialize`.
+  * You can "clone reflectively" a `Box<dyn Reflect>` by using
+    `Reflect::proxy_clone`. It returns a `ReflectProxy`.
+  * You can use `ReflectProxy::sidecast` to create a `Box<dyn Reflect>` from
+    a `ReflectProxy`. The created trait object's underlying type is a concrete
+    type you can downcast to.
+* `FromReflect` doesn't exist anymore. Instead, `Reflect::downcast` provides
+  now the same guarentees that `FromReflect::from_reflect` had. If you need to
+  get a `T` from a `ReflectProxy`, use a combination of
+  `ReflectProxy::sidecast_type<T>` and `downcast`.
+
+
+## Implementation strategy
+
+The `Reflect` API stays identical to the one we have today, at the exception of
+`clone_value`.
+
+We remove the `Dynamic***: Reflect` implementations and we introduce a
+`ReflectProxy` enum.
+
+```rust
+pub enum ReflectProxy {
+    Struct(DynamicStruct),
+    TupleStruct(DynamicTupleStruct),
+    Tuple(DynamicTuple),
+    List(DynamicList),
+    Map(DynamicMap),
+    AlreadyReflect(Box<dyn Reflect>),
+}
+```
+
+`ReflectProxy` mirrors the `ReflectRef` and `TypeInfo` enums.
+
+`ReflectProxy` has a method to convert the underlying value into the
+value it is proxying. This allows you to get a `Reflect` from a `ReflectProxy`
+while upholding the `Reflect` uniqueness rule.
+
+```rust
+impl ReflectProxy {
+  pub fn type_name(&self) -> &str {
+    match self { /* etc. */ }
+  }
+  pub fn can_sidecast(&self, registry: &TypeRegistry) -> Result<(), SidecastError> {}
+  pub fn sidecast(self, registry: &TypeRegistry)
+    -> Result<Box<dyn Reflect>, SidecastError> {
+    let type_name = self.type_name();
+    let registration = registry.get_with_name(type_name).ok_or(SidecastError::Unregistered)?;
```

Comment:
So there are two failure mode:
1. The type itself is not registered
2. The struct the `Dynamic***` is supposed to represent cannot be constructed from the the fields that were added to it. (for example we have `struct Foo { bar: usize }` and a `DynamicStruct::add_field("baz", Box::new("Hi".to_owned()))`)

In the case of storing `Box<ReflectProxy>` instead of `Box<dyn Reflects` in `Dynamic***` fields, the errors could happen when trying to convert the nested value inside one of the fields.